Add E2E test to automate kfto-sft-llm example

ChughShilpa opened the pull request:
Closes [RHOAIENG-18908](https://issues.redhat.com/browse/RHOAIENG-18908)

## Description
This PR adds an e2e test which automates the https://github.com/opendatahub-io/distributed-workloads/tree/main/examples/kfto-sft-llm 

## How Has This Been Tested?
# Prerequitses:
1. Sufficient worker nodes with NVIDIA GPUs (Ampere-based or newer recommended , E.g A100)
2. A dynamic storage provisioner supporting RWX PVC provisioning

Tested manually by running
```
export NOTEBOOK_USER_NAME=
export NOTEBOOK_USER_TOKEN=
export ODH_NAMESPACE=redhat-ods-applications
export TEST_OUTPUT_DIR=
export NOTEBOOK_IMAGE=quay.io/modh/odh-workbench-jupyter-pytorch-cuda-py311-ubi9@sha256:37fae9bb5b1871d2df0d32dfc2a7c4dc654deb0a4bc9744172f61592ae2ff4f8
export HF_TOKEN=
```
`go test -timeout 30m -v ./tests/kfto -run TestKftoSftLlmLlama3_1_8BInstruct`

## Merge criteria:

- [X] The commits are squashed in a cohesive manner and have meaningful messages.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added an end-to-end integration test for fine-tuning large language models using a notebook-based workflow on Kubernetes/OpenShift, including validation of output and math reasoning.
- **Bug Fixes**
  - Improved test reliability by ensuring notebooks use dedicated persistent storage.
- **Refactor**
  - Updated test utilities and test cases to require an explicit persistent volume claim (PVC) for notebooks, enhancing storage handling and resource management.
- **Chores**
  - Corrected environment variable naming in example instructions for HuggingFace token usage.
  - Minor variable naming and formatting updates in example notebooks for clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sutaakar reviewed (commented):
/lgtm

ChughShilpa commented:
> @ChughShilpa I was able to run this test in our team's IBM cluster where PyTorchJob completed successfully but test.Eventually polling logic kept waiting for marker file to be available and then further papermill excecution failed with timeout error.. 👀 I think that the marker file is only created if papermill succeeds. Since papermill is failing, the marker file never gets created. Could you please check this once?
> 
> ```
> kfto_sft_llm_test.go:153: Attempt to read marker file failed: exit status 1
> kfto_sft_llm_test.go:154: Command output: cat: /opt/app-root/src/notebook_completion_marker: No such file or directory
> ```
> 
> ![Screenshot 2025-06-13 at 2 07 03 PM](https://private-user-images.githubusercontent.com/84722973/454790067-1cff226d-8f37-4aa9-9119-b1c1247582ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAwNjY2OTIsIm5iZiI6MTc1MDA2NjM5MiwicGF0aCI6Ii84NDcyMjk3My80NTQ3OTAwNjctMWNmZjIyNmQtOGYzNy00YWE5LTkxMTktYjFjMTI0NzU4MmVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjE2VDA5MzMxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxN2I4NGYwNjFhZjZkMTQ5YmQ0NDNiMzk3OGExZGUyNGNmMDZmZmFhN2RkZWUwMjI4YmQzMWM5ZjhlM2ZhNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.LKkfAmZmDwBX39F-p5PZGItjZHf9hU22v6aRd6BxGI4) ![Screenshot 2025-06-13 at 2 51 56 PM](https://private-user-images.githubusercontent.com/84722973/454790987-15cb81bc-f313-43a3-be75-11e894ce1712.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAwNjY2OTIsIm5iZiI6MTc1MDA2NjM5MiwicGF0aCI6Ii84NDcyMjk3My80NTQ3OTA5ODctMTVjYjgxYmMtZjMxMy00M2EzLWJlNzUtMTFlODk0Y2UxNzEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjE2VDA5MzMxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzZjA0N2E4YjRmYWNkZjU3ZDA4MGQzNGEyZDVmMDQ5ZjViNzc3YzcyNjU4NDdjZjY1MWRiODU0YWJlYmZjMWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ol4s6uyuH7PA_A_Bv82evK6yGu7dkZf07cAAP3NcLww)

Test passed locally at my end.
Make sure you are using the correct notebook image
![image](https://github.com/user-attachments/assets/4c9f618f-cf4d-442a-b090-09e3401a8a7e)

abhijeet-dhumal commented:
/lgtm
Thanks @ChughShilpa !

astefanutti commented on tests/kfto/kfto_sft_llm_test.go:
```diff
@@ -0,0 +1,227 @@
+		test.Eventually(ListNotebooks(test, namespace), TestTimeoutGpuProvisioning).Should(HaveLen(0))
+	}()
+
+	print("Wait for pytorch job to start running ............\n")
```
```suggestion
	test.T().Logf("Wait for pytorch job to start running ...")
```

astefanutti commented on tests/kfto/kfto_sft_llm_test.go:
```diff
@@ -0,0 +1,227 @@
+		Should(WithTransform(PyTorchJobConditionRunning, Equal(corev1.ConditionTrue)))
+	test.T().Logf("PytorchJob sft-%s is running", namespace.Name)
+
+	print("Wait for pytorch job to complete ............\n")
```
```suggestion
	test.T().Logf("Wait for pytorch job to complete ...")
```

astefanutti commented on tests/kfto/kfto_sft_llm_test.go:
```diff
@@ -0,0 +1,227 @@
+		Should(WithTransform(PyTorchJobConditionSucceeded, Equal(corev1.ConditionTrue)))
+	test.T().Logf("PytorchJob sft-%s is successfully completed", namespace.Name)
+
+	print("Wait for notebook execution to complete ............\n")
```
```suggestion
	test.T().Logf("Wait for notebook execution to complete ...")
```

ChughShilpa commented:
@astefanutti @abhijeet-dhumal Please review the latest commit

Refactored papermill execution to write an explicit FAILURE marker in case of failure of notebook execution and added the exit code so the test fails fast and to avoid the timeout error

abhijeet-dhumal commented:
/lgtm

astefanutti commented:
/approve

Thanks for this awesome work @ChughShilpa!
